### PR TITLE
HHH-15933 broken SQL generated for @ManyToOne with @JoinColumn which references a column of a @SecondaryTable

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-15933

### CODE REVIEW

**Comment by gavinking** [comment]
Thanks for working on this!

**Comment by gavinking** [comment]
@mbladel were you still waiting for a review on this or something .... not sure I'm qualified to do it myself.

Perhaps @dreab8 could take a look?

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java [hunk 1]
```diff
@@ -145,7 +147,7 @@ public static void createSyntheticPropertyReference(
 			// figure out which table has the columns by looking
 			// for a PersistentClass or Join in the hierarchy of
 			// the target entity which has the first column
-			final Object columnOwner = findReferencedColumnOwner( targetEntity, joinColumns.getJoinColumns().get(0), context );
+			final AttributeContainer columnOwner = findReferencedColumnOwner( targetEntity, joinColumns.getJoinColumns().get(0), context );
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:150]: Ah, this is a good change. Thanks for that.

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java [hunk 2]
```diff
@@ -334,7 +336,12 @@ private static Property makeSyntheticComponentProperty(
 		result.setInsertable( false );
 		result.setValue( embeddedComponent );
 		result.setPropertyAccessorName( "embedded" );
-		ownerEntity.addProperty( result );
+		if ( ownerEntity instanceof JoinedSubclass ) {
+			ownerEntity.addProperty( result );
+		}
+		else {
+			persistentClassOrJoin.addProperty( result );
+		}
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: Sure, this is actually the one change I was not 100% sure about. I noticed the object where the synthetic `Component` property for the `JoinColumn` was added was often wrong when considering where the referenced column was actually located.

In most simple cases, the column owner (`persistentClassOrJoin`) and `ownerEntity` were the same so this effectively changes nothing.

Even when the column owner and the entity were different, though, the property was always added to the entity. This was causing the column to be always associated with the primary table, and in cases like `SecondaryTable` joins we had no way of knowing the column was actually referencing another table. This change was needed to later be able to get the correct table reference in `ToOneAttributeMapping#createTableGroupJoin`.

The only case where `persistentClassOrJoin` and `ownerEntity` were different and the referenced column was actually in the `ownerEntity`'s side I encountered when inheritance with `InheritanceType.JOINED` was involved, thus the `instanceof` check to add the property to the correct persistent class.

[the remark this answers is not in the document]
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: It's a bit hard to follow all that without some example mappings to see what the individual cases are.

But it seems like if I stipulate that what you're saying is correct, the logic there is still a bit weird.

Wouldn't it be more natural to write:

```java
if ( persistentClassOrJoin instanceof PersistentClass ) {
 	ownerEntity.addProperty( result );
 }
 else { //must be a join
 	persistentClassOrJoin.addProperty( result );
 }
```

But I'm not sure.

It sounds to me like maybe we need additional tests to exercise all the cases you ran into.

[the remark this answers is not in the document]
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: Actually maybe you can just give me an example of the mapping which breaks *without* this change?

[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: probably it should be
```
if ( persistentClassOrJoin instanceof Join ) {
	persistentClassOrJoin.addProperty( result );
}
else {
	ownerEntity.addProperty( result );
}
```

[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: @gavinking 
> Actually maybe you can just give me an example of the mapping which breaks _without_ this change?

Well it doesn't really _break_, but `RefToSecondaryTableTest` mapping of the `Split` entity is the perfect example: without this change the synthetic property is added to the primary table (`Split`) while the referenced column is actually in the secondary one `secondary_split`.

@dreab8 
> what about ...

That should be the same as the new check Gavin proposed, logically. Both look way better than my solution and to me they make sense AFAIU. I'm gonna try to apply them and tests to see what happens.

[the remark this answers is not in the document]
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: > Well it doesn't really break, but `RefToSecondaryTableTest`'s mapping of the `Split` entity is the perfect example: without this change the synthetic property is added to the primary table (`Split`) while the referenced column is actually in the secondary one `secondary_split`.

Right but that example has nothing to do with `InheritanceType.JOINED`, so I'm not seeing where `JoinedSubclass` enters into this.

[the remark this answers is not in the document]
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: (In fact that example has no inheritance at all.)

[the remark this answers is not in the document]
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:344]: > Sorry the inheritance example I encountered was `ManyToManyTest` using `InspectorPrefixes` which is a joined sublcass of `Inspector`.

Ugh. I hate that mess of a package.

Do you think you could create a simpler `RefToJoinedSubclassTest` test case, similar to `RefToSecondaryTableTest`, which demonstrates what breaks when we _don't_ make this change?

Because if there's a problem with references to joined subclasses, then that's actually I guess a *separate bug*, not the same bug I reported.

[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:345]: There was no problem with references to joined subclasses. I just needed to add the synthetic property to the `persistentClassOrJoin`, but for joined subclasses this was wrong so I needed the check.

Anyway I've added the new test to be sure and I rewrote the `if` condition like @dreab8 proposed.

[the remark this answers is not in the document]
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/BinderHelper.java:345]: Ah, sure, sure, now with the updated condition, it's much clearer, actually. I think we're good.

Thank you!!

**Review by beikov — APPROVED** [review]
LGTM

**Comment by gavinking** [comment]
Folks, can't we just merge this now?
